core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+import alluxio.resource.ResourcePool;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Class for managing block read/write locks. After obtaining a lock with
+ * {@link ResourcePool#acquire()}, {@link ResourcePool#release(Object)} must be called when the
+ * thread is done using the client.
+ */
+@ThreadSafe
+public final class BlockRWLockPool extends ResourcePool<ClientRWLock> {
+  private static final Logger LOG = LoggerFactory.getLogger(BlockRWLockPool.class);
+
+  @VisibleForTesting
+  final AtomicBoolean mHasReachedFullCapacity;
+  private final AtomicInteger mRemainingPoolResources;
+
+  /**
+   * Creates a new client read/write lock pool.
+   * @param maxCapacity the max capacity of the pool
+   */
+  public BlockRWLockPool(int maxCapacity) {
+    super(maxCapacity);
+    mHasReachedFullCapacity = new AtomicBoolean(false);
+    mRemainingPoolResources = new AtomicInteger(maxCapacity);
+
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.WORKER_BLOCK_LOCK_POOL_REMAINING_RESOURCES_KEY.toString(),
+        this::getRemainingPoolResources
```

Comment:
does "mRemainingPoolResources::get" work here?

Reply:
I'm just thinking if we could just watch the size of `mResources` from the super class?

Reply:
1. yes it worked. made the change already.
2. unfortunately that doesn't work. Basically the pool creates resources by lazy and the size of mResource is only the number of "created" resources. For example the pool size is 10 and a thread acquires a lock and releases it later. the size of mResource is 1 but the size of remaining pool resources should be 10.